drake/solvers/SNOPTWrapper.cpp

```diff
@@ -27,12 +27,15 @@ using namespace Drake;
 
 // NOTE: all snopt calls will use this shared memory... so this code is NOT THREAD SAFE
 const Drake::OptimizationProblem* current_problem = NULL;
```

Comment:
Yes, and there is a better way to do it.  I had a prototype approach based on this commit that I decided to hold until this is merged.  I'll open a separate PR for it.  Basically we can pass the `OptimizationProblem` pointer through the user data space reserved by `snOptA` for us.